### PR TITLE
Add WhereAbouts IP reconciler support

### PR DESCRIPTION
**Pull request by rollandf**
- Update whereabouts image version to v0.5.1 in helm and doc
- Update whereabouts CRD ippools
- Add whereabouts CRD overlappingrangeipreservations
- Add IP reconciler job yaml

Closes: Issue #298

Signed-off-by: Fred Rolland <frolland@nvidia.com>

### CODE REVIEW

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+          containers:
+            - name: whereabouts
+              image: {{ .CrSpec.Repository }}/{{ .CrSpec.Image }}:{{ .CrSpec.Version }}
+              resources:
+                requests:
+                  cpu: "100m"
+                  memory: "50Mi"
+              command:
+                - /ip-reconciler
+                - -log-level=verbose
```
Comment (e0ne) [inline manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml:45]: Do we really need verbose log level here?

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  successfulJobsHistoryLimit: 0
+  schedule: "*/5 * * * *"
+  jobTemplate:
+    spec:
+      backoffLimit: 0
+      template:
+        metadata:
+          labels:
+            app: whereabouts
+        spec:
```
Comment (e0ne) [inline manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml:33]: Please, add imagePullSecrect as we did for whereabouts [1]


[1] https://github.com/Mellanox/network-operator/blob/7da462f954c3160f41f5b543318bfb77f099f98b/manifests/stage-whereabouts-cni/0050-whereabouts-ds.yaml#L51

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: ip-reconciler
```
Comment (e0ne) [inline manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml:17]: nit: please rename it to 'whereabouts-ip-reconciler'

**Review by e0ne — CHANGES_REQUESTED** [review]

**Comment by rollandf** [comment]
With local env run:
```
# kubectl -n nvidia-network-operator-resources get jobs
NAME                     COMPLETIONS   DURATION   AGE
ip-reconciler-27403820   0/1           2m10s      2m10s

```

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml (hunk 2 above)
Comment (rollandf) [reply manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml:33]: Done

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml (hunk 1 above)
Comment (rollandf) [reply manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml]: changed to `info`

File: manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml (hunk 3 above)
Comment (rollandf) [reply manifests/stage-whereabouts-cni/0060-ip-reconciler-job.yaml]: done

**Comment by rollandf** [comment]
Logs from Job Pod:

```
# kubectl -n nvidia-network-operator-resources logs whereabouts-ip-reconciler-27405230-25j4q
2022-02-08T09:50:01Z [debug] NewReconcileLooper - Kubernetes config file located at: 
2022-02-08T09:50:01Z [debug] listing IP pools
2022-02-08T09:50:01Z [debug] the IP reservation: IP: 192.168.2.225 is reserved for pod: default/rdma-test-pod-1
2022-02-08T09:50:01Z [debug] pod ref default/rdma-test-pod-1 is not listed in the live pods list
2022-02-08T09:50:01Z [debug] pod ref default/rdma-test-pod-1 is not listed in the live pods list
2022-02-08T09:50:01Z [debug] Going to update the reserve list to: []
2022-02-08T09:50:01Z [debug] successfully cleanup IPs: [192.168.2.225]
2022-02-08T09:50:01Z [verbose] removed stale overlappingIP allocation [192.168.2.225]
```

**Review by e0ne — COMMENTED** [review]
Please, add a note into the commit message and PR description:

Closes: Issue #298

**Comment by adrianchiris** [comment]
@AbdYsn seems like CI is stuck, do you know why ?

**Comment by AbdYsn** [comment]
> @AbdYsn seems like CI is stuck, do you know why ?

@adrianchiris I am trying to debug the issue with the setup.
when an sriov network node policy is created the setup gets rebooted, causing the jobs to fail.
working on debugging and fixing the issue .

**Comment by rollandf** [comment]
> > @AbdYsn seems like CI is stuck, do you know why ?
> 
> @adrianchiris I am trying to debug the issue with the setup. when an sriov network node policy is created the setup gets rebooted, causing the jobs to fail. working on debugging and fixing the issue .

Hi, @AbdYsn any update?

File: deploy/role.yaml [hunk 4]
```diff
@@ -291,3 +292,32 @@ rules:
...
+  resources:
+  - cronjobs
+  verbs:
+  - create
+  - delete
+  - get
+  - list
+  - patch
+  - update
+  - watch
```
Comment (adrianchiris) [inline deploy/role.yaml:323]: nit: newline

File: manifests/stage-whereabouts-cni/0020-cluster_role.yaml [hunk 5]
```diff
@@ -28,3 +29,20 @@ rules:
...
+  - get
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups: [""]
+  resources:
+  - pods
+  verbs:
+  - list
```
Comment (adrianchiris) [inline manifests/stage-whereabouts-cni/0020-cluster_role.yaml:48]: nit: newline

**Review by adrianchiris — COMMENTED** [review]
two very minor nits, otherwise LGTM

**Review by adrianchiris — APPROVED** [review]
LGTM, just need CI to pass

**Review by adrianchiris — APPROVED** [review]

**Pull request merged**